File: student-next/src/types/lesson.ts
export interface LessonSummary {
  id: number;
  module_id: number;
  title: string;
  order: number;
  status: string;
  description?: string | null;
  difficulty?: string | null;
  estimated_time?: number | null;
  language?: string | null;
  blocks_order?: number[] | null;
}

export interface LessonBlock {
  id?: number;
  order?: number;
  block_type?: string;
  blockType?: string;
  type?: string;
  title?: string;
  word?: string;
  content?: string;
  data?: any;
  payload?: any;
  url?: string;
  video_url?: string;
  image_url?: string;
  audio_url?: string;
  markdown?: string;
  word_id?: number;
  word_ids?: number[];
  cards?: Flashcard[];
  items?: Flashcard[];
  question?: string;
  choices?: string[];
  options?: string[];
  answer?: number;
  transcript?: string;
  translation?: string;
  expected_pronunciation?: string;
  phrase?: string;
  sample_audio_url?: string;
}

export interface Flashcard {
  id?: number;
  front?: string;
  back?: string;
  word?: string;
  translation: string;
  image_url?: string | null;
  audio_url?: string | null;
  example_sentence?: string;
}

export interface Quiz {
  id?: number;
  question: string;
  options: string[];
  correct_option: number;
  explanation?: string | null;
  order?: number;
}

export type BlockType =
  | "video"
  | "theory"
  | "audio_theory"
  | "image"
  | "audio"
  | "flashcards"
  | "pronunciation"
  | "quiz"
  | "theory_quiz"
  | "lesson_test"
  | "audio_task";

export type TheoryContent = {
  title?: string;
  markdown?: string;
  rich_text?: string;
  video_url?: string;
  thumbnail_url?: string;
};

export type FlashcardsContent = {
  cards: Flashcard[];
};

export type PronunciationItem = {
  id?: number | string;
  word?: string;
  translation?: string;
  example?: string;
  image_url?: string | null;
  audio_url?: string | null;
};

export type PronunciationContent = {
  items: PronunciationItem[];
  sample_audio_url?: string;
  expected_pronunciation?: string;
  phrase?: string;
};

export type TaskQuestion = {
  question: string;
  type: "audio" | "audio_repeat" | "single" | "single_choice" | "fill-in" | "open";
  options?: string[];
  audio_url?: string;
  placeholder?: string;
};

export type TasksContent = {
  questions: TaskQuestion[];
};

export interface LessonDetail {
  lesson: LessonSummary & {
    module: {
      id: number;
      order?: number;
      course?: { slug?: string };
    };
    blocks?: LessonBlock[];
    flashcards?: Flashcard[];
    quizzes?: Quiz[];
  };
  blocks: LessonBlock[];
  flashcards: Flashcard[];
  quizzes: Quiz[];
  progress_status: string;
  score?: number | null;
  time_spent?: number | null;
  course_progress: number;
  module_progress: number;
  progress_map: Record<number, string>;
  navigation: {
    prev_lesson_id?: number | null;
    next_lesson_id?: number | null;
  };
  new_words_added?: number;
}

File: backend/app/schemas/lesson.py
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field

from .block import LessonBlockOut, NormalizedBlock


class FlashcardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    front: str
    back: str
    image_url: Optional[str] = None
    audio_path: Optional[str] = None
    audio_url: Optional[str] = None
    order: int


class QuizOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    question: str
    options: list[str]
    correct_option: int
    explanation: Optional[str] = None
    order: int


class LessonSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    module_id: int
    title: str
    order: int
    status: str
    difficulty: Optional[str] = None
    estimated_time: Optional[int] = None
    language: Optional[str] = None
    version: Optional[int] = None
    blocks_order: List[int] | None = None
    is_deleted: bool = False
    description: Optional[str] = None
    video_type: Optional[str] = None
    video_url: Optional[str] = None


class LessonOut(LessonSummary):
    blocks: List[LessonBlockOut] = Field(default_factory=list)


class LessonNavigation(BaseModel):
    prev_lesson_id: Optional[int] = None
    next_lesson_id: Optional[int] = None


class LessonDetail(BaseModel):
    lesson: LessonSummary
    blocks: List[NormalizedBlock] = Field(default_factory=list)
    flashcards: List[FlashcardOut] = Field(default_factory=list)
    quizzes: List[QuizOut] = Field(default_factory=list)
    progress_status: str
    score: Optional[int] = None
    time_spent: Optional[int] = None
    course_progress: int
    module_progress: int
    progress_map: dict
    navigation: LessonNavigation


class LessonCreate(BaseModel):
    module_id: int
    title: str
    description: Optional[str] = None
    status: str = "draft"
    difficulty: Optional[str] = None
    estimated_time: Optional[int] = None
    age_group: Optional[str] = None
    language: str = "kk"
    order: Optional[int] = None
    video_type: Optional[str] = "youtube"  # youtube, vimeo, file
    video_url: Optional[str] = None


class LessonUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    difficulty: Optional[str] = None
    estimated_time: Optional[int] = None
    age_group: Optional[str] = None
    language: Optional[str] = None
    version: Optional[int] = None
    blocks_order: Optional[List[int]] = None
    order: Optional[int] = None
    video_type: Optional[str] = None  # youtube, vimeo, file
    video_url: Optional[str] = None


__all__ = [
    "LessonSummary",
    "LessonOut",
    "LessonDetail",
    "FlashcardOut",
    "QuizOut",
    "LessonNavigation",
    "LessonCreate",
    "LessonUpdate",
]
